### PR TITLE
Clipspace implementations for left/righted-handed coordinate systems and [-1,1] and [0,1] clipspace

### PR DESCRIPTION
**Pull request by raedwulf**

### CODE REVIEW

**Comment by raedwulf** [comment]
@michael-g @recp The tests are still failing but it's a start for some early reviewing

File: CMakeLists.txt
Comment (michael-g) [inline CMakeLists.txt:61]: This duplicates the line above
```suggestion
  src/vec2.c
```

File: include/cglm/call/clipspace/ortho_lh_no.h
Comment (michael-g) [inline include/cglm/call/clipspace/ortho_lh_no.h:20]: With sincere apologies for being a terrible pedant, and please feel free to ignore, when renaming the parameter to `nearZ`, the following whitespace has not increased to align the `float farZ` parameter (here and elsewhere -- let me know and I can suggest changes to fix).
```suggestion
                 float bottom,  float top,
                 float nearZ,   float farZ,
```

Comment (raedwulf) [reply include/cglm/call/clipspace/ortho_lh_no.h:20]: No worries! Please let me know where the whitespace is wrong!

File: include/cglm/call/clipspace/ortho_lh_zo.h
Comment (michael-g) [inline include/cglm/call/clipspace/ortho_lh_zo.h:20]:
```suggestion
                 float bottom,  float top,
                 float nearZ,   float farZ,
```

File: include/cglm/call/clipspace/ortho_rh_no.h
Comment (michael-g) [inline include/cglm/call/clipspace/ortho_rh_no.h:20]:
```suggestion
                 float bottom,  float top,
                 float nearZ,   float farZ,
```

File: include/cglm/call/clipspace/ortho_rh_zo.h
Comment (michael-g) [inline include/cglm/call/clipspace/ortho_rh_zo.h:20]:
```suggestion
                 float bottom,  float top,
                 float nearZ,   float farZ,
```

File: include/cglm/call/clipspace/persp_lh_no.h
Comment (michael-g) [inline include/cglm/call/clipspace/persp_lh_no.h:20]:
```suggestion
                   float bottom,  float top,
                   float nearZ,   float farZ,
```

File: include/cglm/call/clipspace/persp_lh_zo.h
Comment (michael-g) [inline include/cglm/call/clipspace/persp_lh_zo.h:20]:
```suggestion
                   float bottom,  float top,
                   float nearZ,   float farZ,
```

File: include/cglm/call/clipspace/persp_rh_no.h
Comment (michael-g) [inline include/cglm/call/clipspace/persp_rh_no.h:20]:
```suggestion
                   float bottom,  float top,
                   float nearZ,   float farZ,
```

File: include/cglm/call/clipspace/persp_rh_zo.h
Comment (michael-g) [inline include/cglm/call/clipspace/persp_rh_zo.h:20]:
```suggestion
                   float bottom,  float top,
                   float nearZ,   float farZ,
```

File: include/cglm/cam.h
Comment (michael-g) [inline include/cglm/cam.h:351]: Where are `CGLM_CLIP_CONTROL_LH` and `CGLM_CLIP_CONTROL_RH` defined?

Comment (raedwulf) [reply include/cglm/cam.h]: Good catch! was meant to be CGLM_CLIP_CONTROL_LH_BIT/RH_BIT - incidentally the tests pass now!

**Comment by michael-g** [comment]
Will hopefully do some more later on, but this is a huge amount of work you've done, @raedwulf. Awesome stuff.

**Comment by recp** [comment]
Hi @raedwulf, @michael-g,

Sorry for the late response :/

Many thanks for this work, as @michael-g said this is a huge work.

Should we separate files by `persp_[lh|rh]_[no|zo].h`, `ortho_[lh|rh]_[no|zo].h` ... or `proj_[lh|rh]_[no|zo].h` and `view_[lh|rh]_[no|zo].h` ?

Also there is a struct api, we should add wrappers fo this change into `include/cglm/stuct/clipspace`. Struct is just wrappers to existing array api.

**Comment by raedwulf** [comment]
I'll look into the struct API. Currently we do have an extra separation for view too. Would proj just be the combination of ortho and persp? @recp

**Comment by recp** [comment]
> Would proj just be the combination of ortho and persp?

Yes, but I'm not sure which way is better, separating them or joining them. There are lot of persp functions and maybe come more for both ortho and persp, in this case separating them may be good idea.

File: include/cglm/common.h
Comment (recp) [inline include/cglm/common.h:69]: Can we add these options to `docs/source/opt.rst`? People can recognize the options quickly when reading docs.

**Comment by recp** [comment]
Maybe this is not one of goal of the PR and out of scope, but I'm wonder that should **cglm** need anything else to support LH. For instance transforms (matrix rotations, quaternions, vector rotations...).

Also there are frustum helpers in `frustum.h` e.g extracts planes, corners... Unproject functions in `project.h`... I'm not sure what else should be done to sully support LH.

Actually uses should achieve same rotation around X and Y by negating angle but Z should remain same I think. Maybe we should discuss these in another issue[s].

**Comment by recp** [comment]
@raedwulf ping. 

I think we need to add tests for other functions too, after that we can merge this PR if there are no any issues

**Comment by michael-g** [comment]
Do you have a sense of how you'd like to approach the testing? We could continue with the small amount of work I've done and paste a load of magic numbers into the tests, or go ahead with the new repository which compares cglm implementations again the glm library? The former is much more of a unit test, obviously, but the latter has its own benefits.

**Comment by raedwulf** [comment]
I looked into the struct API and I can get it done soonish :) thanks @michael-g that would be great!

**Comment by recp** [comment]
> Do you have a sense of how you'd like to approach the testing? We could continue with the small amount of work I've done and paste a load of magic numbers into the tests, or go ahead with the new repository which compares cglm implementations again the glm library? The former is much more of a unit test, obviously, but the latter has its own benefits.

@michael-g I think we should add unit tests first then library-to-library test can be written later. The magic numbers could come from the math instead of glm directly, or both maybe, glm also needs tests... I have just created a repo called [cglm-libtests](https://github.com/recp/cglm-libtests), the name can be changed with better one...

> I looked into the struct API and I can get it done soonish :)

Awesome!

**Comment by recp** [comment]
It would be nice to finish this work asap, I can merge this PR and we can add tests with new PRs maybe? 

@raedwulf @michael-g any progress?

**Comment by raedwulf** [comment]
@recp Just a bit busy last few days, but I'll spend some time this weekend to finalise things with the struct api.

**Comment by recp** [comment]
@raedwulf great 🤗

**Comment by michael-g** [comment]
> It would be nice to finish this work asap, I can merge this PR and we can add tests with new PRs maybe?
> 
> @raedwulf @michael-g any progress?

Sorry, I've been busy with work and family stuff. 

If only we could auto-gen a bunch of tests using results from GLM...

**Comment by raedwulf** [comment]
I pushed the struct api stuff @recp @michael-g

**Comment by recp** [comment]
@raedwulf awesome work, 

There was no need to add clip control to default strruct api like this:

```C
CGLM_INLINE
mat4s
glms_frustum(float left,   float right,
             float bottom, float top,
             float nearZ,  float farZ) {
#if CGLM_CONFIG_CLIP_CONTROL == CGLM_CLIP_CONTROL_LH_ZO
  return glms_frustum_lh_zo(left, right, bottom, top, nearZ, farZ);
#elif CGLM_CONFIG_CLIP_CONTROL == CGLM_CLIP_CONTROL_LH_NO
  return glms_frustum_lh_no(left, right, bottom, top, nearZ, farZ);
#elif CGLM_CONFIG_CLIP_CONTROL == CGLM_CLIP_CONTROL_RH_ZO
  return glms_frustum_rh_zo(left, right, bottom, top, nearZ, farZ);
#elif CGLM_CONFIG_CLIP_CONTROL == CGLM_CLIP_CONTROL_RH_NO
  return glms_frustum_rh_no(left, right, bottom, top, nearZ, farZ);
#endif
}
```

clip control already done in `glm_frustum()` this could be just a wrapper as before. But since you did it, let's keep it.

**Comment by recp** [comment]
Now we need to add tests ;)

**Review by recp — APPROVED** [review]

**Comment by raedwulf** [comment]
Ahh, yeah that's true. That file could be reverted with no problems though.

**Comment by recp** [comment]
@raedwulf @michael-g I'm going to merge this PR, but let's improve this by adding tests, bug fixes with new PRs...

This is still valid question for me: https://github.com/recp/cglm/pull/198#issuecomment-841684755

Feel free to open issues and bring tests, bugfixes
 
@raedwulf @michael-g many thanks for this work 🤗

**Comment by recp** [comment]
@raedwulf @michael-g the PR is merged, many thanks 🤗 🎉

Let's write tests ;)

**Comment by michael-g** [comment]
@raedwulf, kudos, amazing work. Will try to contribute this week to the testing code. Again: is there a source of reference implementations?

**Comment by raedwulf** [comment]
Hey @michael-g, they are based on the cglm source code and the glm code for the differences.

**Comment by warmwaffles** [comment]
@raedwulf Love this! Thank you so much for pushing this through.

**Comment by raedwulf** [comment]
@warmwaffles thanks! We're still in the process of testing this and any feedback will be great!

**Comment by recp** [comment]
We should announce this feature after the tests are done e.g. reddit, gamedev... I'll try what I can... After used by community then more improvements can be made. 

Great to have great community 🎉

**Comment by quadroli** [comment]
Wow, been following along this project for a while, glad to see it come to fruition, excellent work all at @michael-g @raedwulf and @recp for the insights, kudos 🥳

**Comment by quadroli** [comment]
Quick question, are there noticeable to changes on how to use the library now?

**Comment by recp** [comment]
@quadroli thanks, 

> are there noticeable to changes on how to use the library now

Actually not, just optionally it is possible to use other clipspaces rather than OpenGL's one now, if there are no any bugs

**Comment by quadroli** [comment]
Awesome 👍
Love the new look to, iss clean 🧼

**Pull request merged**
